### DISCUSSION TITLE
Cross platform GPU, compute and draw maybe.

### DISCUSSION THREAD

**Discussion by maybeJosiah** [community/community#146421] [Programming Help]
### Body

Okay, I am trying to use GPU stuff on all major platforms, (Linux, Mac, Windows, (maybe BSDs), Nintendo Switch, Xbox, PlayStation, iOS, and iPadOS, Steam Deck, maybe Apple TVOS), and finding I may need some transpiler or cross platform thing or just to write in multiple programming languages. Going for a semi-native thing. I would need compute shaders and drawing shaders. So far I have seen HLSL (compatible with Windows and Xbox), GLSL (Vulkan and OpenGL, supports most computers, Nintendo Switch, and can be installeded on PlayStation) MSL (For all Apple stuff mentioned), NVN (Nintendo Switch), PSSL (PlayStation), and I do not know what BSDs take. Looks like if I do this correct, GLSL, HLSL, and MSL should cover everything. Is there a better way than write for all these? Bonus, WebGL uses GLSL so web support too. There is also WGSL for web. I hear BSDs use OpenCL. Steam Deck has Vulkan drivers so GLSL works for that. CUDA is for Nvidia GPUs.

I would also need to use all this stuff somehow with CPU stuff. I prefer Zig, Rust, or Go, would be okay using Kotlin, Swift, or JavaScript too. Any suggestions? I have also heard of Slang (it is a shaders language). I am trying to build a 4d+ GPU graphics system. Do I need to build an engine entirely myself including GPU stuff or can I use something existing? Is there possibly an automated translator where I could get raw language output? Any ideas appreciated. Looks like this will be multi-backend, possibly single front-end with multiple things that could run based on system. I prefer like a write once, run anywhere thing to some capacity but okay if I must write 3+ shader files.

Where to even start? GPUs are notoriously good for graphics and especially intensive graphics. Compute shaders are for from 4d+ points to 2d screen points and what order to draw them in. Also, as performant and native as possible is preferred. I am still developing anyway. I am amazing at code translation so not much hassle going to many shaders. Compiled is preferred. JIT is okay. Bytecode is less okay. X E.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by maybeJosiah** [comment]
According to like my research, Spir-V is now supported on Direct3d for Xbox and Windows, Spir-V works in Vulkan, and there is an official library to translate Spir-V to MSL for Apple, it looks like anything that can make Spir-V can be used and that is at least GLSL and HLSL. Android supports Vulkan. It looks like just choose a language, use Spir-V, and it can run on about everything. BSDs can run OpenGL so GLSL seems like a good start choice. X E.

**Comment by AliAli2320** [comment]
- [ ]

**Reply by maybeJosiah** [reply]
What?

**Comment by Vedant0109** [comment]
Oh
